rts/lra/lra-coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/Transaction.java

```diff
@@ -389,9 +391,10 @@ private void updateAfterLRAListeners(RecordList list) {
         while ((r = i.iterate()) != null) {
             if (r instanceof LRARecord) {
                 URI endNotification = ((LRARecord) r).getEndNotificationUri();
+                URI recoveryCoordinatorURI = ((LRARecord) r).getRecoveryCoordinatorURI();
 
-                if (endNotification != null) {
-                    afterLRAListeners.add(endNotification);
+                if (endNotification != null && recoveryCoordinatorURI != null) {
```

Comment:
Is it ever possible for recoveryCoordinatorURI to be == null but endNotification be != null. Is that covered by spec as an error?

Reply:
No, neither of them can be null. I did the check just to be ultra cautious. The background to the two fields is:

A class annotated with both `@AfterLRA` and `@LRA` will receive LRA end notifications if the `@LRA` method is invoked. When the method is invoked we enlist the entity with our coordinator by creating an instance of LRARecord (https://github.com/jbosstm/narayana/blob/master/rts/lra/lra-coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/Transaction.java#L552). The LRARecord constructor initialises the endNotification URI. Immediately after construction [we set the recoveryCoordinatorURI](https://github.com/jbosstm/narayana/blob/master/rts/lra/lra-coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/Transaction.java#L557) (I did it that way because it is easier to get the arjuna uid of the record after creation and I didn't see a huge benefit in generating the recovery URI inside the constructor). But it does mean that initialisation the two fields happens in different methods but if there was a crash in between then nothing would make it to disk. Therefore the null check is redundant but I regard it as a harmless check (if I didn't have the check I guess I'd have to document in the code why they cannot both be null).

Reply:
@tomjenkinson actually I remember why I don't initialise the recovery URI in the LRARecord constructor: it is because I need to know the JAX-RS base URL of the server that generates the recovery URI and the class that constructs the LRARecord knows what that is.